Lib/fontTools/pens/basePen.py

```diff
@@ -131,6 +132,19 @@ def addComponent(
         """
         raise NotImplementedError
 
+    def addVarComponent(
+        self,
+        glyphName: str,
+        transformation: VarTransform,
+        location: Dict[str, float],
+    ) -> None:
+        """Add a VarComponent sub glyph. The 'transformation' argument
+        must be a VarTransform from the fontTools.misc.transform module,
+        and the 'location' argument must be a dictionary mapping axis tags
+        to their locations.
+        """
+        raise NotImplementedError
```

Comment:
IIUC, in the ttGlyphSet.draw/drawPoints you are interpreting NotImplementedError exception as a signal that the given pen cannot handle drawing varComponents and thus one wishes to "decompose" them. However, wouldn't it be better to use AttributeError (or hasattr(pen, "addVarComponent") instead of NotImplementedError? The latter usually is used for astract classes where the expectation is that all concrete subclasses _must_ implement the abstract method, otherwise the concrete class cannot be instantiated. Whereas in your case, this addVarComponet seems to be intentionally designed to be "optionally implementable".
An optional method with that interface looks more suitable for this maybe. So if a pen wants to handle them varComponents it can define such a method, otherwise they will be drawn decomposed.

Reply:
or... could the whole decomposing logic be moved to the basePen.DecomposingPen (or another similar mixin), the one that takes a glyphSet to resolve references? Then you would no longer catch NotImplementedError from within TTGlyphSet.draw, but simply call the pen's addVarComponent method and leave it to the pen itself to do the decomposing (or not do it)... I don't know what's better, honestly.

Reply:
Moving the decomposing logic to the pen is hard because we need to keep more state (the font's location at least) that the pen doesn't have. That's why currently the VarComposite pens are drawn in ttGlyphSet.

Reply:
I mean, I can make basePen.DecomposingPen call a private method of ttGlyphPen...

Reply:
> An optional method with that interface looks more suitable for this maybe. So if a pen wants to handle them varComponents it can define such a method, otherwise they will be drawn decomposed.

So, you suggest basePen wouldn't have this method at all? I find that clumsy.

Reply:
> you suggest basePen wouldn't have this method at all?

yeah that'd also be weird. What I was a bit concerned about was that defining an addVarComponent on AbstractBasePen in the same way as the other abstract methods (i.e. return NotImplementedError) conveys the idea that any concrete class that inherit from them ought to implement it, whereas in this case it looks like a pen may deliberately "decide" to leave it unimplemented in order to trigger the ttGlyphSet decomposing behavior..

Reply:
> As you said addVarComposite is an opt-in, not mandatory method now.

to me raising NotImplementedError communicates a sense of urgency, "implement me!". However, now this seems to be part of the interface itself, a way for a pen to deliberately shift the responsibility of decomposing varcomponents to the "context" (the DecomposingRecordingPen overrides the RecordingPen.addVarComponents by re-raising NotImplementedError exactly to communicate to TTGlyphSet that it indeed wants them to be decomposed).
Maybe this is ok and I am overthinking this. Maybe it's ok that for the time being only TTGlyphSet knows what varComponents are and can decompose them.

Reply:
Would `raise AttributeError` would look nicer to you?

Reply:
With a comment?

Reply:
hm.. I didn't think of that, actually. Yeah, it kind of communicates better the idea (even if it's technically lying). It's like the equivalent of not defining the method at all (which is the norm for all existing pens), but communicates that it is a possible optional method with given signature that one may want to implement in order to handle varcomponents (unless they want them decomposed). Good idea!